### ISSUE TITLE
Allow Read{Only,Write}Paths= to apply to symlink inodes themselves

### ISSUE THREAD

**Issue by moocan** [systemd/systemd#19320]
**Is your feature request related to a problem? Please describe.**
It seems to be impossible to access or modify a file without exposing the whole directory content.
We have a use case where we need to (as example):
- One path unit is monitoring a file and start a service when this file is updated.

The service unit need to
- Update (when required) content of a file in `/run/app/file.conf` through a `RuntimeDirectory=` inside the service unit which is symlinked inside `/etc/app.conf` (no choice) and sent to bash script with `RUNTIME_DIRECTORY` variable
- Touch the symlink `/etc/app.conf` to update the time stamp or create the symlink when not existing

We tried to restrict acces to only this `file/symlink` in `/etc` using `ProtectSystem=strict` and `ReadWritePaths=/etc/app.conf`, `ReadOnlyPaths=/etc`, `TemporaryFileSystem=/etc`, `BindPaths`, `BindReadOnly` directives but can't find a way to perform the `touch` or `ln` without exposing the whole `/etc/**/*` content to the bash script. 

Also we tried to restrict access on some files using `InaccessiblePaths` as for example `InaccessiblePaths=/etc/passwd` but this seems to have no effect with a `ReadWritePaths=/etc` (which is required to be allowed to touch or create the symlink) because the bash script is accessing these files even with an `InaccessiblePaths`. 

**Describe the solution you'd like**
A way to access or modify file/symlink using parent drectory, file system permissions without exposing the whole content of a folder, a sort of whitelisting only some file/folder inside a blacklisted paths while honoring permissions.

**Describe alternatives you've considered**
To be honest, we don't know if it's possible or if it's only related to our case or just a bad service unit configuration because we found some information about `TemporaryFileSystem` or `Read{Only,Write}Paths` but always with folder path but never  with directly a file/symlink.

**The systemd version you checked that didn't have the feature you are asking for**
We are testing with systemd version 245.

Regards

**Comment by poettering** [comment]
Not sure I follow, but TemporaryFileSystem=/tmp + ReadWritePaths=/etc/app/file.conf should work, no?

**Comment by moocan** [comment]
@poettering 

Hello,
Thank you for your return.

The app `file.conf`(on which ne we need to touch and ln) is not inside a subfolder in `/etc` as `/etc/app/file.conf` but (and to our great misfortune) is directly inside `/etc` as `/etc/file.conf`

so, Yes and No.
Using these settings

> ProtectSystem=true
> ProtectHome=true
> ReadOnlyPaths=/
> ReadWritePaths=/etc
> TemporaryFileSystem=/etc
> InaccessiblePaths=/etc/passwd

`touch` or `ln` `/etc/file.conf -> /run/app/file.conf` are working (this is what we are required to do after upadting /run/app/file.conf contents) but bash script can access all files in /etc/ as `/etc/passwd` or edit/create new file anywhere in `/etc/**/*/` (which is problematic)

Using these settings

> ProtectSystem=true
> ProtectHome=true
> ReadOnlyPaths=/
> ReadWritePaths=/etc/file.conf
> TemporaryFileSystem=/etc
> InaccessiblePaths=/etc/passwd

`touch` or `ln` `/etc/file.conf -> /run/app/file.conf` are not working anymore (read-only file system)  but bash script can access all files in /etc/ as `/etc/passwd` but can't no more edit/create new file anywhere in /etc/**/*/ (which is already a good thing but so we can't do some of what we need to do on /etc/file.conf)

Regards

**Comment by poettering** [comment]
I don't understand a word, this is very confusingly written. Please try again, and say what you want to do, and do *not* say how you tried to fix it. The fact you mix up both makes this really confusing to me.

**Comment by moocan** [comment]
@poettering 

Sorry.
Whe have an application which need to access a file to read it's configuration
- this file is `/etc/appsettings.conf` inside `/etc`
- the file `/etc/appsettings.conf` is a symlink to `/run/appname/settings.conf`
- the file `/run/appname/settings.conf`is updated with a systemd path unit which is monitoring another file for updates

To force the application to read its settings again, we need to
- update the time stamp of symlink `touch -h /etc/appsettings.conf`

If this symlink does not exist or is invalid (eg at startup or later for any reason) 
- create the symlink with `ln -s [-f] /run/appname/settings.conf /etc/appsettings.conf`

All steps on the file/symlink `/etc/appsettings.conf` are performed by a bash script which is launched by a systemd .service unit. 

Regards

**Comment by poettering** [comment]
Read{Only,Write}Paths= and InaccessiblePaths= expose Linux mount concepts as unit file settings. On Linux, you cannot make symlinks mount points (if you try, the symlink will be followed, and whatever it points to is mounted, and if you pin the inode via O_PATH, then you'll get EPERM if you try to overmount it). Thus, these two options only apply to symlinks through the next upper mount, they cannot apply just to the symlink itself.

As I understand this is what you are looking for, right? You want the symlink that is located immediately inside of /etc to be there, but you don't want the rest of /etc to be there. Linux simply doesn't allow this to be configured. Sorry.

Or did I misunderstand something?

**Comment by moocan** [comment]
> As I understand this is what you are looking for, right? You want the symlink that is located immediately inside of /etc to be there, but you don't want the rest of /etc to be there. Linux simply doesn't allow this to be configured. Sorry.
> 
> Or did I misunderstand something?

You are right. 
The application leaves us no choice.

I found an issue / RFE on being able to not follow the symbolic link and use it as if it were a file without following it (but I can't find it anymore) 

I was hoping for something like 
```
InaccessiblePaths=/
ReadWritePaths=/etc/appsettings.conf
```
as a way to "hide" or "filter" the contents of /etc to a single file/symlink  to restrict bash script access to this target only without accessing other files or folders. But maybe to easy :)

As a workaround I have created a working AppArmor profile and integrated it to the service unit.
To avoid any risk, all access by the bash script to anything other than /etc/appsettings.conf is now denied. 

Thank you for taking the time and for your explanations.
Regards
